File: backend/PhoneStore.Domain/Entities/Product.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace PhoneStore.Domain.Entities;

public class Product
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonElement("name")]
    public string Name { get; set; } = null!;

    [BsonElement("description")]
    public string Description { get; set; } = string.Empty;

    [BsonElement("price")]
    [BsonRepresentation(BsonType.Decimal128)]
    public decimal Price { get; set; }

    [BsonElement("salePrice")]
    [BsonRepresentation(BsonType.Decimal128)]
    public decimal SalePrice { get; set; }

    [BsonElement("brandId")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string BrandId { get; set; } = null!;

    [BsonElement("categoryId")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string CategoryId { get; set; } = null!;

    [BsonElement("images")]
    public List<string> Images { get; set; } = new();

    [BsonElement("specifications")]
    public ProductSpecification Specifications { get; set; } = new();

    [BsonElement("stock")]
    public int Stock { get; set; }

    [BsonElement("sold")]
    public int Sold { get; set; }

    [BsonElement("averageRating")]
    public double AverageRating { get; set; }

    [BsonElement("totalReviews")]
    public int TotalReviews { get; set; }

    [BsonElement("isActive")]
    public bool IsActive { get; set; } = true;

    [BsonElement("createdAt")]
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [BsonElement("updatedAt")]
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public class ProductSpecification
{
    [BsonElement("ram")]
    public string Ram { get; set; } = string.Empty;

    [BsonElement("rom")]
    public string Rom { get; set; } = string.Empty;

    [BsonElement("screenSize")]
    public string ScreenSize { get; set; } = string.Empty;

    [BsonElement("battery")]
    public string Battery { get; set; } = string.Empty;

    [BsonElement("cpu")]
    public string Cpu { get; set; } = string.Empty;

    [BsonElement("os")]
    public string Os { get; set; } = string.Empty;

    [BsonElement("color")]
    public string Color { get; set; } = string.Empty;
}

File: mobile/lib/core/models/product.dart
class Product {
  final String id;
  final String name;
  final String description;
  final double price;
  final double salePrice;
  final String brandId;
  final String brandName;
  final String categoryId;
  final String categoryName;
  final List<String> images;
  final int stock;
  final int sold;
  final double averageRating;
  final int totalReviews;
  final List<String> promotions;
  final List<ProductStorageVariant> storageVariants;
  final List<ProductColorVariant> colorVariants;
  final ProductSpec specifications;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    required this.salePrice,
    required this.brandId,
    required this.brandName,
    required this.categoryId,
    required this.categoryName,
    required this.images,
    required this.stock,
    required this.sold,
    required this.averageRating,
    required this.totalReviews,
    required this.promotions,
    required this.storageVariants,
    required this.colorVariants,
    required this.specifications,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    var storageList = json['storageVariants'] as List? ?? [];
    var colorList = json['colorVariants'] as List? ?? [];
    
    return Product(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      salePrice: (json['salePrice'] as num?)?.toDouble() ?? 0.0,
      brandId: json['brandId'] ?? '',
      brandName: json['brandName'] ?? '',
      categoryId: json['categoryId'] ?? '',
      categoryName: json['categoryName'] ?? '',
      images: List<String>.from(json['images'] ?? []),
      stock: json['stock'] ?? 0,
      sold: json['sold'] ?? 0,
      averageRating: (json['averageRating'] as num?)?.toDouble() ?? 0.0,
      totalReviews: json['totalReviews'] ?? 0,
      promotions: List<String>.from(json['promotions'] ?? []),
      storageVariants: storageList.map((x) => ProductStorageVariant.fromJson(x)).toList(),
      colorVariants: colorList.map((x) => ProductColorVariant.fromJson(x)).toList(),
      specifications: ProductSpec.fromJson(json['specifications'] ?? {}),
    );
  }
}

class ProductStorageVariant {
  final String storage;
  final double price;
  final double salePrice;

  ProductStorageVariant({
    required this.storage,
    required this.price,
    required this.salePrice,
  });

  factory ProductStorageVariant.fromJson(Map<String, dynamic> json) {
    return ProductStorageVariant(
      storage: json['storage'] ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      salePrice: (json['salePrice'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class ProductColorVariant {
  final String name;
  final String imageUrl;
  final double priceModifier;

  ProductColorVariant({
    required this.name,
    required this.imageUrl,
    required this.priceModifier,
  });

  factory ProductColorVariant.fromJson(Map<String, dynamic> json) {
    return ProductColorVariant(
      name: json['name'] ?? '',
      imageUrl: json['imageUrl'] ?? '',
      priceModifier: (json['priceModifier'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class ProductSpec {
  final String ram;
  final String rom;
  final String screenSize;
  final String battery;
  final String cpu;
  final String os;
  final String color;

  ProductSpec({
    required this.ram,
    required this.rom,
    required this.screenSize,
    required this.battery,
    required this.cpu,
    required this.os,
    required this.color,
  });

  factory ProductSpec.fromJson(Map<String, dynamic> json) {
    return ProductSpec(
      ram: json['ram'] ?? '',
      rom: json['rom'] ?? '',
      screenSize: json['screenSize'] ?? '',
      battery: json['battery'] ?? '',
      cpu: json['cpu'] ?? '',
      os: json['os'] ?? '',
      color: json['color'] ?? '',
    );
  }
}
